[Inductor] Fix assertion failure in efficient_conv_bn_eval for default args

dumko2001 opened the pull request:
Summary:
The [efficient_conv_bn_eval](cci:1://file:///Users/ioop/Documents/github/pytorch/torch/_inductor/fx_passes/efficient_conv_bn_eval.py:17:0-75:17) pass previously asserted that `batch_norm` nodes have exactly 8 arguments. This caused a crash when `batch_norm` was called with default arguments (e.g., only 3 positional args), which is valid in FX graphs.
This PR replaces the brittle assertion with `inspect.signature` binding to correctly normalize arguments and fill defaults, ensuring the pass works for all valid `batch_norm` calls.
Test Plan:
- Reproduced the crash with a script using `F.batch_norm(input, mean, var)` (3 args).
- Verified the fix with the same script (now passes).
- Ran existing tests: `python test/inductor/test_efficient_conv_bn_eval.py` (passed).
Fixes #169011

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @azahed98 @mlazos

dumko2001 committed:
Fix assertion failure in efficient_conv_bn_eval for default args

jansel reviewed (changes requested):
Could you add a test for what this is fixing?

dumko2001 committed:
Add regression test for issue #169011

dumko2001 committed:
Fix Pyrefly lint error: assert callable(bn_node.target)

jansel reviewed (changes requested):
Fix failing lints/tests

dumko2001 committed:
Fix lint error: replace assert with safety check in efficient_conv_bn_eval

dumko2001 committed:
Address jansel review: use local bound_args instead of mutating node

- Remove duplicate @inductor_config.patch decorator in test
- Remove commented assertion and unnecessary callable check
- Use local normalized_args variable instead of mutating bn_node.args
- Add docstring to satisfy linter

jansel commented:
@pytorchbot merge

dumko2001 committed:
Fix PYFMT lint: add blank line between imports

eellison reviewed (commented):
looks good, but can you add the test from original issue as well?

```
from torch.fx import Graph, GraphModule
import torch

from torch._inductor import config as inductor_config

inductor_config.efficient_conv_bn_eval_fx_passes = True
inductor_config.freezing = False   # just to be safe for this pass

graph = Graph()

# Create input placeholders
inp = graph.placeholder("input")
mean = graph.placeholder("mean")
var = graph.placeholder("var")

# Create a call_function node for addition

z = graph.call_function(torch.nn.functional.batch_norm, args = (inp, mean, var))

# Create output node
graph.output(z)

# Wrap in a GraphModule
gm = GraphModule({}, graph)
print(gm)

pycode = graph.python_code('self')
print("Transformed python code: ")
print(pycode.src)

gm_compiled = torch.compile(gm, backend = 'inductor')

inp = torch.randn(4, 4)   # or cpu
mean = torch.randn(4, )
var = torch.randn(4, )

out = gm_compiled(inp, mean, var)
print(out)
```

dumko2001 committed:
Add FX graph regression test for issue #169011

dumko2001 commented:
@eellison Added  test_fx_graph_batch_norm_defaults
 which uses direct FX Graph construction with torch.nn.functional.batch_norm(inp, mean, var)

mlazos commented:
@pytorchbot merge -i

dumko2001 commented:
Rebased onto current upstream main.

The auto-revert on Mar 6 was triggered because the detectron2 graph break counts in the CSV files were off by 1 — the fix causes `efficient_conv_bn_eval` to skip optimization for those models (since their `batch_norm` calls use default args), adding one graph break each for `maskrcnn_r_101_fpn`, `maskrcnn_r_50_c4`, and `maskrcnn_r_50_fpn`. The last commit updates those expected counts to match.

dumko2001 commented on benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_amp_freezing_torchbench_inference.csv:
```diff
@@ -82,15 +82,15 @@ detectron2_maskrcnn_r_101_c4,fail_accuracy,56
 
 
-detectron2_maskrcnn_r_101_fpn,fail_accuracy,62
+detectron2_maskrcnn_r_101_fpn,fail_accuracy,63
```
totally my bad , asked an AI assistant to help rebase it, and I din watch over it , and it incorrectly updated the CSV values to match a past CI run thinking that would fix things.

jansel commented:
@claude what is going on with that CI failure?  Is it related?

jansel commented:
@pytorchbot rebase

jansel commented:
@claude are those CI failures related?

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge

jansel commented:
@claude explain those CI failures, are they related?

jansel commented:
@pytorchbot rebase

jansel commented:
@claude explain those CI failures, are they related?

jansel commented:
@pytorchbot rebase

jansel commented:
@claude explain those CI failures, are they related?

jansel commented:
@pytorchbot rebase

jansel commented:
@claude explain those CI failures, are they related?

jansel commented:
@pytorchbot rebase

jansel commented:
@claude are those failures related?

jansel commented:
@pytorchbot merge -i
